src/tactic/positivity.lean

```diff
@@ -0,0 +1,343 @@
+/-
+Copyright (c) 2022 Mario Carneiro, Heather Macbeth. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Mario Carneiro, Heather Macbeth
+-/
+import tactic.norm_num
+
+/-! # `positivity` tactic
+
+The `positivity` tactic in this file solves goals of the form `0 ≤ x` and `0 < x`.  The tactic works
+recursively according to the syntax of the expression `x`.  For example, a goal of the form
+`0 ≤ 3 * a ^ 2 + b * c` can be solved either
+* by a hypothesis such as `5 ≤ 3 * a ^ 2 + b * c` which directly implies the nonegativity of
+  `3 * a ^ 2 + b * c`; or,
+* by the application of the lemma `add_nonneg` and the success of the `positivity` tactic on the two
+  sub-expressions `3 * a ^ 2` and `b * c`.
+
+For each supported operation, one must write a small tactic, tagged with the attribute
+`@[positivity]`, which operates only on goals whose leading function application is that operation.
+Typically, this small tactic will run the full `positivity` tactic on one or more of the function's
+arguments (which is where the recursion comes in), and if successful will combine this with an
+appropriate lemma to give positivity of the full expression.
+
+This file contains the core `positivity` logic and the small tactics handling the basic operations:
+`min`, `max`, `+`, `*`, `/`, `⁻¹`, raising to natural powers, and taking absolute values.  Further
+extensions, e.g. to handle `real.sqrt` and norms, can be found in the files of the library which
+introduce these operations.
+
+## Main declarations
+
+* `tactic.positivity.base` is the base case of the recursion
+* `tactic.positivity.attr` creates the `positivity` user attribute for tagging the extension tactics
+  handling specific operations, and specifies the behaviour for a single step of the recursion
+* `tactic.positivity.core` collects the list of tactics with the `@[positivity]` attribute and calls
+  the first recursion step as specified in `tactic.positivity.attr`.  Its input is `e : expr` and
+  its output (if it succeeds) is a pair `(bool, expr)`, with the `expr` a proof of the
+  strict-positivity/nonnegativity of `e` and the `bool` indicating whether what could be proved was
+  strict-positivity or nonnegativity
+* `tactic.interactive.positivity` is the user-facing tactic.  It parses the goal and, if it is of
+  one of the forms `0 ≤ e`, `0 < e`, `e > 0`, `e ≥ 0`, it sends `e` to `tactic.positivity.core`.
+
+## TODO
+
+Implement extensions for other operations (raising to non-numeral powers, `exp`, `log`, coercions
+from `ℕ` and `ℝ≥0`).
+
+-/
+
+namespace tactic
+
+/-- Given two tactics whose result is `bool × expr`, report a `bool × expr`:
+- if at least one gives `tt`, report `tt` and one of the `tt` `expr`s
+- if neither gives `tt` but at least one gives `ff`, report `ff` and one of the `ff` `expr`s
+- if both fail, fail -/
+meta def orelse' (tac1 tac2 : tactic (bool × expr)) : tactic (bool × expr) := do
```

Comment:
For clarity, can we replace this `bool` with
```lean
inductive rel_type
| positivity
| non_negativity
```
or similar?

Reply:
Following a suggestion of Mario I switched to
```lean
meta inductive strictness : Type
| positive : expr → strictness
| nonnegative : expr → strictness
```
Do you like it?  I do!

Reply:
The name feels a bit weird now with that definition, but otherwise I think it's fine